astro/src/content/docs/extend/code/lambdas/testing.mdx

```diff
@@ -0,0 +1,429 @@
+    console.info('Goodbye World!');
+  }
+```
+
+Save the file and upload the lambda with the following command.
+
+```bash
+npx fusionauth lambda:update f3b3b547-7754-452d-8729-21b50d111505 --key lambda_testing_key
+```
+
+You can check that the lambda in FusionAuth now says "Goodbye World!" by viewing the ["[ATest]" lambda details](http://localhost:9011/admin/lambda).
+
+<img src="/img/docs/extend/code/lambdas/updated-lambda.png" alt="Update lambda" role="bottom-cropped" width="1200" />
+
+### CLI Limitations
+
+The Node CLI allows you only to create, retrieve, and update lambdas. You can delete a lambda that is not in use by an application with `lambda:delete`. The way to link or unlink a lambda with an application is through the admin UI, API, or a client library.
+
+For example, to link a lambda with an application in the TypeScript client library, you could use code similar to the following.
+
+```ts
+const request: ApplicationRequest = {
+  application: {
+    lambdaConfiguration: {
+      accessTokenPopulateId: "f3b3b547-7754-452d-8729-21b50d111505"
+    }
+  }
+};
+await new FusionAuthClient(apiKey, host).patchApplication(applicationId, request);
+```
+
+## Testing Overview
+
+Lambdas run arbitrary code at certain points in an authentication flow. For example, you can use lambdas to:
+
+- Get user information from an external provider (like a first name or photo from Facebook) or the User object in FusionAuth to put in a JWT.
+- Call an external service at login, for example, to send any suspicious login attempt to a private Slack channel monitored by administrators.
+
+In both these cases, there are two types of tests you can perform:
+
+- **Integration test**: Check if the lambda has uploaded and is running correctly by logging in and seeing if the expected output happens.
+- **Unit test:** You don't upload the lambda, but instead create a mock FusionAuth event that calls the lambda in your code and checks that the lambda does what it is supposed to.
+
+Each of these types of lambda tests is outlined below.
+
+### Test Library
+
+There are many JavaScript test libraries available, and everyone has their preference. This guide uses a simple library for demonstration so that you can generalize the tests to your favorite library. The tests below use [`tape`](https://github.com/ljharb/tape), which implements the [Test Anything Protocol (TAP)](https://en.wikipedia.org/wiki/Test_Anything_Protocol), a language-agnostic specification for running tests that's been in use since 1987. The tests also use [fetch-mock](https://www.wheresrhys.co.uk/fetch-mock/) to mock `fetch` calls from your lambda.
+
+Install the following in your test app terminal.
+
+```bash
+npm install --save-dev tape
+npm install --save-dev faucet # a little test-runner to give neat tape output
+npm install --save-dev fetch-mock
+npm install --save-dev jsonwebtoken # to decode the JWT
+npm install --save-dev uuid # to make a random user Id
+```
+
+<Aside type="note">
```

Comment:
do we want to update to work with modern Node.js too too?

Reply:
This is actually related to [`fusionauth-typescript-client`](https://github.com/FusionAuth/fusionauth-typescript-client), so we'd have to upgrade it there.